balancer.go

```diff
@@ -0,0 +1,256 @@
+package xlb
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/xdire/xlb/tlsutil"
+	"net"
+	"strings"
+	"sync"
+)
+
+type Options struct {
+	Logger   Logger
+	LogLevel string
+}
+
+type LoadBalancer struct {
+	id           string
+	runCtx       context.Context
+	killCtx      context.CancelFunc
+	logger       Logger
+	poolMap      map[string]ServicePool
+	forwarderMap map[string]*Forwarder
+	mutex        sync.Mutex
+}
+
+func NewLoadBalancer(ctx context.Context, cfgPool []ServicePool, opt Options) (*LoadBalancer, error) {
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return nil, fmt.Errorf("gen id failed")
+	}
+	if ctx == nil {
+		return nil, fmt.Errorf("missing parameter ctx")
+	}
+	if cfgPool == nil || len(cfgPool) == 0 {
+		return nil, fmt.Errorf("missing parameter cfgPool")
+	}
+
+	logger := opt.Logger
+	if logger == nil {
+		logger = newZeroLogForName("xlb", id.String(), opt.LogLevel)
+	}
+
+	poolMap := make(map[string]ServicePool)
+	for _, pool := range cfgPool {
+		if len(pool.Identity()) == 0 {
+			return nil, fmt.Errorf("pool missing identity")
+		}
+		poolMap[pool.Identity()] = pool
+	}
+
+	derCtx, cancelFunc := context.WithCancel(ctx)
+	return &LoadBalancer{
+		id:           id.String(),
+		runCtx:       derCtx,
+		killCtx:      cancelFunc,
+		logger:       logger,
+		forwarderMap: map[string]*Forwarder{},
+		poolMap:      poolMap,
+	}, nil
+}
+
+func (lb *LoadBalancer) UpdatePool(pool ServicePool) error {
+	if len(pool.Identity()) == 0 {
+		return fmt.Errorf("pool missing identity")
+	}
+	lb.mutex.Lock()
+	defer lb.mutex.Unlock()
+	lb.poolMap[pool.Identity()] = pool
+	if fwd, exists := lb.forwarderMap[pool.Identity()]; exists {
+		fwd.UpdateServicePool(pool)
+	}
+	return nil
+}
+
+// Listen will try to launch balancer on all the required ports, strategy is all or nothing
+func (lb *LoadBalancer) Listen() error {
+
+	// Do this step to ensure that we will fail on misconfiguration if more than
+	// one service pool mapping presented to this load balance for mTLS
+	// TODO: TLS Allows manual verification for the handshake by that we can launch multiple pools on the same port
+	// TODO: for the first version we will limit 1 pool per port for simplicity
+	mapping, err := collectListenTargets(lb.poolMap)
+	if err != nil {
+		return err
+	}
+
+	type schedule struct {
+		port int
+		tls  *tls.Config
+		pool ServicePool
+	}
+	scheduleListeners := make([]schedule, 0)
```

Comment:
Why not use the known capacity?
```suggestion
	scheduleListeners := make([]schedule, len(mapping))
```

Reply:
It was due to the following reasons:
- no expectations for performance at this particular stage
- no additional index to add as code written to traverse the map

Decided not to do premature optimization. 
Will be fixed.